# Which csv?

## database.csv or fatalencounters??? or both??

We are going to try to see if we can use the name as a temporary key to mix the two databases

In [18]:
# Importing pandas

import pandas as pd
import datetime as dt

In [19]:
# Importing the 2 databases

wpost = pd.read_csv ('/Users/ReDi/Documents/GitHub/data-analysis-project/data/database.csv')
fenc = pd.read_csv ('/Users/ReDi/Documents/GitHub/data-analysis-project/data/fatalencounters.csv')

In [20]:
# Let´s have a quick look at the 2 databases.
# The Washington Post one:

wpost.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2142 entries, 0 to 2141
Data columns (total 14 columns):
id                         2142 non-null int64
name                       2123 non-null object
date                       2142 non-null object
manner_of_death            2142 non-null object
armed                      2136 non-null object
age                        2099 non-null float64
gender                     2141 non-null object
race                       2039 non-null object
city                       2142 non-null object
state                      2142 non-null object
signs_of_mental_illness    2142 non-null bool
threat_level               2142 non-null object
flee                       2106 non-null object
body_camera                2142 non-null bool
dtypes: bool(2), float64(1), int64(1), object(10)
memory usage: 205.1+ KB


In [21]:
# And now the Fatal encounters one:

fenc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20792 entries, 0 to 20791
Data columns (total 24 columns):
Timestamp                                                                          20791 non-null float64
Subject's name                                                                     20792 non-null object
Subject's age                                                                      20207 non-null object
Subject's gender                                                                   20757 non-null object
Subject's race                                                                     20791 non-null object
URL of image of deceased                                                           7084 non-null object
Date of injury resulting in death (month/day/year)                                 20792 non-null object
Location of injury (address)                                                       20315 non-null object
Location of death (city)                           

At first glance it can appear the second one to be more in everything, more entries, more columns. The first thing is true: we have indeed more entries...but...more columns?

Name/ age / gender / race, are common to both databases ... so 4 - 4

talking about place: Wpost has city / state
                     Fenc has address/city/state/zip/county
                                                        ... so 6 - 9  (3 columns of additional info about location in fenc)
                                                        
We have date, manner (or cause) of death and mental illness in both..9 - 12

We have 4 unique columns in wpost:
Armed / Threat level / Flee / Body Camera                   ...13 - 12  ( 4 columns exclusive of wpost)

We have several unique columns in fenc, but some expendable:
URL of image of deceased / A brief description of the circumstances surrounding the death / Link to news article or photo of official document / State Data Status / Date&Description / Date (Year) / 20791 are expendable (we can keep the description for curiosity purposes)
We have 2 identifiers: de unique id and the timestamp ( we just need one)
We have 2 uniqe columns: agency and official dispotition    ...so 13 - 14  (2 uniques in fenc)


We have 5 columns unique in the fenc (3 of them about the location), and 4 unique in the wpost

In [22]:
# Comparing the 2 databases entries (just a remembrance)

wpost.shape, fenc.shape

((2142, 14), (20792, 24))

### Time frame of each database

Let´s see, more or less, wich is the difference of entries between our 2 databases considering a similar time frame.

In [23]:
fenc['Date (Year)'].min(), fenc['Date (Year)'].max()

(2000, 2100)

In [24]:
# Since we have a problem here, let´s see what is going to happen in the 2100


fenc[fenc['Date (Year)'] > 2017]

Timestamp                                    Subject's name  \
20791        NaN  Items below this row have not been fact-checked.   

      Subject's age Subject's gender Subject's race URL of image of deceased  \
20791           NaN              NaN            NaN                      NaN   

      Date of injury resulting in death (month/day/year)  \
20791                                  December 31, 2100   

                            Location of injury (address)  \
20791  Items below this line have not been fact-checked.   

      Location of death (city) Location of death (state)     ...      \
20791                      NaN                       NaN     ...       

      A brief description of the circumstances surrounding the death  \
20791                                                NaN               

      Official disposition of death (justified or other)  \
20791                                                NaN   

      Link to news article or photo of official document  \
20791                     http://www.fatalencounters.org   

      Symptoms of mental illness? Unique identifier/submitted by  \
20791                         NaN                     Bill jones   

      State Data Status Prior to Jan. 1, 2013. All states complete after Jan. 1, 2013  \
20791                                                  X                                

                                  Date&Description    20791 Unique identifier  \
20791  12/31/2100:  http://www.fatalencounters.org  20791.0               NaN   

      Date (Year)  
20791        2100  

[1 rows x 24 columns]

In [25]:
# So, we have like and ending entrie that has no value to our database, le´s get rid of it

fenc = fenc.drop(20791)

In [26]:
fenc['Date (Year)'].min(), fenc['Date (Year)'].max()

(2000, 2017)

So our fatal encounters database goes from 2000 to 2017. We have solved the error of the 2100 entry dropping it.

In [27]:
# Let´s see the time frame of wpost, and the number of entries

wpost.date.min(), wpost.date.max(), wpost.id.count()

('2015-01-02', '2017-02-28', 2142)

In [28]:
# Approx, in the same time frame, in the fenc:

fenc[fenc['Date (Year)'] >= 2015].shape

(3919, 24)

We can see here that approximately, the fatal encounters db is doble the size of the washington post one. Let's try to be more accurate.

Since our column about the date of the killing in the fenc database is a string, and not a date, we cannot use it. We have to transform the data. In order to do that e are going to use the datetime library. We are going to use it to give format of a date to this column. We are going to create a new column where we are going to have this date data in a proper way.



In [30]:
# We iterate through every row, using iterrows. First we create our new column, with a string type

for i, row in fenc.iterrows():

    # we define n, as our string to be transformed for every entry
    n = row['Date of injury resulting in death (month/day/year)']

    # and here we define our new column, date, that is going to be n transformed into date. For doing that, we have to
    # describe which is the original format. In our case the date is written on a string like 'January 23, 2005'.
    # If we go to the documentation of the datetime library, we see that %B stands for the complete name of the month,
    # %d stands for the day number, then a coma, and the the year in 4 digits format.
    # The date in the end is to have just the date, without hours minutes and seconds
    fenc['date'].iloc[i] = dt.datetime.strptime(n,'%B %d, %Y')

C:\Users\ReDi\Anaconda3\lib\site-packages\pandas\core\indexing.py:141: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [33]:
fenc[(fenc['date'] >= '2015-01-02') & (fenc['date'] <= '2017-02-28')].shape

(3420, 25)

So in the end, for the same time frame the washington post dataframe is 2142 entries, the fatal encounters one is 3420. Can we try to merge them? Since the second one is 1300 entries bigger, we can asume that every name in the wpost is indeed in the fatal encounters one...but we can have typos, names that don´t match, missing names...let´s investigate.


## Merging the csv's

Just for future uses, and because my computer is very slow, let's save this as a new csv

In [38]:
fenc.to_csv('fenc.csv', index = False, encoding='utf-8')

Timestamp               Subject's name Subject's age Subject's gender  \
0            1.0        LaTanya Janelle McCoy            24           Female   
1            2.0                Lester Miller            53             Male   
2        19512.0             James Martin Loy            55             Male   
3            4.0          John Edward Pittman            45             Male   
4            5.0             John Frank Brown            20             Male   
5            3.0              Derrick E. Tate            23             Male   
6            8.0                 Adrian Dolby            31             Male   
7            7.0           Darryl Woodall Jr.            23             Male   
8            6.0                  Kyle Dillon            19             Male   
9           10.0               Joseph Gumpert            28             Male   
10           9.0      Name withheld by police            28             Male   
11          13.0            Sonny Daniel Diaz            24             Male   
12          11.0                  Lamon Davis           NaN             Male   
13          12.0             Benjamin K. Olds            20             Male   
14          14.0           Erin Dudley Forbes            26             Male   
15       19347.0             Harold Greenwald            41             Male   
16       19348.0                  Paul Carson            68             Male   
17          15.0              Guido Paul Jean            17             Male   
18          16.0          Dennis Carl Johnson            49             Male   
19          17.0             Jorge Montelongo           NaN             Male   
20          19.0          Donald Dean Compton            44             Male   
21          18.0           Jamie Fowler Boyll            50           Female   
22          20.0        Miguel Peralta Torres            41             Male   
23          21.0            Robert H. Golding            43             Male   
24          22.0         Jose Daniel Gonzalez            25             Male   
25       19349.0           George Blair Reese            38             Male   
26          24.0              Kortique Walter            17             Male   
27          23.0          Daniel Millan Lopez            21             Male   
28          25.0                Alan Zelencic            28             Male   
29          26.0             Snapper Mitchell           NaN             Male   
...          ...                          ...           ...              ...   
20761    20642.0      Name withheld by police           NaN             Male   
20762    20643.0        Matthew Colton Stover            21             Male   
20763    20737.0         Mark Raymond Tillman            52             Male   
20764    20644.0                 Michael Bell            37             Male   
20765    20645.0      Name withheld by police            28             Male   
20766    20738.0                  David Pacas            36             Male   
20767    20739.0                 Robb Stewart            50             Male   
20768    20740.0                Pedro Ramirez            28             Male   
20769    20741.0               Jacob Faulkner            32             Male   
20770    20742.0           David Bryan Creson            40             Male   
20771    20743.0         Brandon D. Lukenbill            33             Male   
20772    20744.0  Daniel Francisco Valenzuela            37             Male   
20773    20745.0          Armando Garcia-Muro            17             Male   
20774    20746.0              Santino Trevino            20             Male   
20775    20747.0          Jesus Ramon Deltoro            34             Male   
20776    20748.0                 Donald Ready            77             Male   
20777    20749.0   Christopher Michael Murray           NaN             Male   
20778    20750.0               Jimmie Bevenue            47             Male   
20779  